tests/test_lab_dev/test_circuit_components.py

```diff
@@ -37,6 +41,79 @@ def test_light_copy(self):
         assert d.y is d_copy.y
         assert d.wires is not d_copy.wires
 
+    def test_matmul_one_mode(self):
+        r"""
+        Tests that ``__matmul__`` produces the correct outputs for one-mode components.
+        """
+        vac0 = Vacuum(1)
+        d0 = Dgate(1, modes=[0])
+        a0 = Attenuator(0.9, modes=[0])
+
+        result1 = vac0 @ d0
+        result1 = (result1 @ result1.adjoint) @ a0
+
+        assert result1.wires == Wires(modes_out_bra=[0], modes_out_ket=[0])
+        assert np.allclose(result1.representation.A, 0)
+        assert np.allclose(result1.representation.b, [0.9486833, 0.9486833])
+        assert np.allclose(result1.representation.c, 0.40656966)
+
+        result2 = result1 @ vac0.dual @ vac0.dual.adjoint
+        assert not result2.wires
+        assert np.allclose(result2.representation.A, 0)
+        assert np.allclose(result2.representation.b, 0)
+        assert np.allclose(result2.representation.c, 0.40656966)
+
+    def test_matmul_multi_modes(self):
+        r"""
+        Tests that ``__matmul__`` produces the correct outputs for multi-mode components.
+        """
+        vac012 = Vacuum(3)
+        d0 = Dgate(0.1, 0.1, modes=[0])
+        d1 = Dgate(0.1, 0.1, modes=[1])
+        d2 = Dgate(0.1, 0.1, modes=[2])
+        a0 = Attenuator(0.8, modes=[0])
+        a1 = Attenuator(0.8, modes=[1])
+        a2 = Attenuator(0.7, modes=[2])
+
+        result = vac012 @ d0 @ d1 @ d2
+        result = result @ result.adjoint @ a0 @ a1 @ a2
+
+        assert result.wires == Wires(modes_out_bra=[0, 1, 2], modes_out_ket=[0, 1, 2])
+        assert np.allclose(result.representation.A, 0)
+        assert np.allclose(
```

Comment:
@sylviemonet @ziofil 
In mrmustard.lab, one would get the same results by doing ``(vac012 >> d0 >> d1 >> d2 >> a0 >> a1 >> a2).bargmann()`. If I do this, I get the Bargmann triple of a pure state. I guess that the attenuator is mapping ket to ket? But in lab_dev, it is a channel

Just wondering if this is the reason

Reply:
Not wrong with Attenuator. 

Just inside bargmann() function, it checks if the state is pure or not and gives the triples. 
In our lab_dev yes, we make the decision on our own not on the state to check at this point.

Reply:
Thanks for checking this!

Here is the other discrepancy, which I find when I project a multi-mode state:

MrMustard lab:
```
import mrmustard.lab as mm

vac0 = mm.Vacuum(1)
vac01 = mm.Vacuum(2)
vac012 = mm.Vacuum(3)
d0 = mm.Dgate(.1, .1, modes=[0])
d1 = mm.Dgate(.1, .1, modes=[1])
d2 = mm.Dgate(.1, .1, modes=[2])
a0 = mm.Attenuator(.8, modes=[0])
a1 = mm.Attenuator(.8, modes=[1])
a2 = mm.Attenuator(.7, modes=[2])

result = vac012 >> d0 >> d1 >> d2 >> a0 >> a1 >> a2 << vac012
result
```


MrMustard lab_dev:
```
import mrmustard.lab_dev as mmd

vac0 = mmd.Vacuum(1)
vac01 = mmd.Vacuum(2)
vac012 = mmd.Vacuum(3)
d0 = mmd.Dgate(.1, .1, modes=[0])
d1 = mmd.Dgate(.1, .1, modes=[1])
d2 = mmd.Dgate(.1, .1, modes=[2])
a0 = mmd.Attenuator(.8, modes=[0])
a1 = mmd.Attenuator(.8, modes=[1])
a2 = mmd.Attenuator(.7, modes=[2])

result = vac012 @ d0 @ d1 @ d2
result = (result.adjoint) @ a0 @ a1 @ a2 @ vac012.dual @ vac012.dual.adjoint
result.representation.c
```

You can observe similar discrepancies if you project onto `vac01`

Reply:
`result = (result @ result.adjoint) @ a0 @ a1 @ a2 @ vac012.dual @ vac012.dual.adjoint` would this be result and result adjoint at the beginning?

Reply:
is this resolved or does it need discussing?

Reply:
Resolved and fixed in one of @sylviemonet's latest PRs